app/models/entry.rb

```diff
@@ -1,4 +1,11 @@
 class Entry < ApplicationRecord
   belongs_to :user
   belongs_to :room
+  include Partner
+
+  # メッセージが１つでも存在する部屋のメッセージを返す処理
```

Comment:
Modelのクラスメソッド で繰り返し処理をしたい場合はfind_eachでも取得できます。selfは省略できる。
```suggestion
def self.sorted_entries(user)
    last_messages = []
    find_each do |another_entry|
    # メッセージが１つでも存在する部屋のメッセージを取得
      if another_entry.room.messages.present?
        last_messages << another_entry.room.messages.last
      end
    end
    # メッセージの作成時刻をもとにソート
    sorted_last_messages = last_messages.sort_by! { |a| a[:created_at] }.reverse
    sorted_entries = []
    sorted_last_messages.each do |sorted_last_message|
      sorted_entries << sorted_last_message.room.entries.partner_of(user)
    end
    sorted_entries
end
```

Reply:
ありがとうございます。
インスタンスメソッドにしていたので動かなかったようです。

私の認識では、
作成したオブジェクトに対して使うメソッド -> インスタンスメソッド
すべてのオブジェクトに共通して使えるメソッド -> クラスメソッド　でした。
そのため、一度作成したオブジェクトを並び替える = インスタンスメソッド
かと思っていたのですが、違うのですね。

クラスメソッドとインスタンスメソッドの違いを詳しく教えていただけないでしょうか。
たとえば、今回の場合ですとなぜクラスメソッドにしたのでしょうか。

Reply:
> クラスメソッドとインスタンスメソッドの違いを詳しく教えていただけないでしょうか。

ちょっと雑な解説ですが、1レコードに対する処理に関してはインスタンスメソッド。複数のレコードの処理に関してはクラスメソッド という理解でいいかと思います。

インスタンスメソッドの例
```ruby
class Entry < ApplicationRecord
  # Entry.first(1).user_name など、複数のレコードの処理をする時に参照する。
  # 親モデルのuserの名前を返すメソッド
  def user_name
    user.name
  end
end
```

クラスメソッドの例
```ruby
class Entry < ApplicationRecord
  # Entry.where(user_id: [1, 2, 3]).user_names など、複数のレコードの処理をする時に参照する。
  # また、Entry.user_names というように呼ぶこともできる。この場合はEntry全件に対する処理となる。
  # 全てのレコードの親モデルのuserの名前を返すメソッド
  def self.user_names
    includes(:user).map {|r|  r.user.name }
  end
end
```

Reply:
> 1レコードに対する処理に関してはインスタンスメソッド。複数のレコードの処理に関してはクラスメソッド という理解でいいかと思います。

ありがとうございます。やっと理解できました。
確かに今回は「ソート＝複数のレコードを扱う」ですので、クラスメソッドですね。